### PR TITLE
Edit & delete icon functionality for document widget

### PR DESCRIPTION
**Pull request by djnunez-aot**
*Issue #881 :*
https://github.com/bcgov/met-public/issues/881

-Add folder & document deletion

-Add folder & document data editing

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the met-public license (Apache 2.0).

### CODE REVIEW

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx [hunk 1]
```diff
@@ -1,26 +1,107 @@
...
+        setEdit(false);
+        if (document.title !== documentItem.title) {
+            const documentUpdatesToPatch = updatedDiff(documentItem, {
+                ...document,
+            }) as PatchDocumentRequest;
+            await patchDocument(documentWidget.id, document.id, {
+                ...documentUpdatesToPatch,
+            });
+
+            loadWidgets();
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx:43]: why load widgets?

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentsContext.tsx [hunk 2]
```diff
@@ -62,17 +73,30 @@ export const DocumentsProvider = ({ children }: { children: JSX.Element | JSX.El
         }
     };
 
-    useEffect(() => {
-        loadDocuments();
-    }, [savedEngagement, widget]);
+    const clearSelected = () => {
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentsContext.tsx:76]: where has this been used?

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx [hunk 3]
```diff
@@ -27,10 +29,13 @@ type FileForm = yup.TypeOf<typeof schema>;
 
 const FileDrawer = () => {
     const dispatch = useAppDispatch();
-    const { documents, handleFileDrawerOpen, fileDrawerOpen, loadDocuments, widget } = useContext(DocumentsContext);
-
+    const { documentToEdit, documents, handleFileDrawerOpen, fileDrawerOpen, loadDocuments, widget } =
+        useContext(DocumentsContext);
+    const { loadWidgets } = useContext(WidgetDrawerContext);
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:34]: loadwidgets never used

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx [hunk 4]
```diff
@@ -53,13 +64,29 @@ const FileDrawer = () => {
...
             setIsCreatingDocument(true);
-            await postDocument(widget.id, {
-                title: data.name,
-                parent_document_id: data.folderId === 0 ? null : data.folderId,
-                url: data.link,
-                widget_id: widget.id,
-                type: 'file',
-            });
+            if (documentToEdit) {
+                const documentEditsToPatch = updatedDiff(documentToEdit, {
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:68]: this is fine, perhaps do this though just to improve readability:

like in EngagementForm.tsx
    const handleSaveEngagement = () => {
        if (isNewEngagement) {
            return handleCreateEngagement();
        }

        return handleUpdateEngagement();
    };

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx [hunk 5]
```diff
@@ -88,7 +115,14 @@ const FileDrawer = () => {
                         padding="2em"
                     >
                         <Grid item xs={12}>
-                            <MetHeader3 bold>Add File</MetHeader3>
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:91]: This is preferred when the condition is for a lot of code, but if it's a simple text perhaps it's just easier to go with your instincts and use ternary here like <MetHeader3> { !!documentToEdit ? 'Edit File' : 'Add File'} </MetHeader3>

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx [hunk 6]
```diff
@@ -140,10 +174,13 @@ const FileDrawer = () => {
                                     InputLabelProps={{
                                         shrink: false,
                                     }}
+                                    defaultValue={parentDocument?.title}
                                     fullWidth
                                     size="small"
                                 >
-                                    <MenuItem key={`folder-option-0`} value={0} sx={{ height: '2em' }}></MenuItem>
+                                    <MenuItem key={`folder-option-0`} value={0} sx={{ height: '2em' }}>
+                                        none
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:182]: kinda liked the empty space, if a 'none' text here is preferred maybe italicize it?

**Review by tiagograf-AOT — APPROVED** [review]

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentsContext.tsx (hunk 2 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentsContext.tsx]: removed

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx (hunk 5 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:91]: good idea

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx (hunk 6 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx]: italicized

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx [hunk 7]
```diff
@@ -1,26 +1,126 @@
...
+    const { documents, loadDocuments } = useContext(DocumentsContext);
+    const { widgets, loadWidgets } = useContext(WidgetDrawerContext);
+    const documentWidget = widgets.find((widget: Widget) => widget.widget_type_id === WidgetType.Document);
+    const [edit, setEdit] = useState<boolean>(false);
+    const [document, setDocument] = useState<DocumentItem | undefined>(documentItem);
+
+    useEffect(() => {
+        const updatedDocument = documents.find((document: DocumentItem) => document.id === documentItem.id);
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // ts-ignore:
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx:31]: what is being ignored here?

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx [hunk 8]
```diff
@@ -42,25 +45,63 @@ const FileDrawer = () => {
...
+            return;
+        }
+        const documentEditsToPatch = updatedDiff(documentToEdit, {
+            title: data.name,
+            parent_document_id: data.folderId === 0 ? null : data.folderId,
+            url: data.link,
+        }) as PatchDocumentRequest;
+        await patchDocument(widget.id, documentToEdit.id, {
+            ...documentEditsToPatch,
+            parent_document_id: data.folderId === 0 ? null : data.folderId,
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:70]: but this way we are always sending parent_document_id, can we have it that we only send it if it changed

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx [hunk 9]
```diff
@@ -42,25 +45,63 @@ const FileDrawer = () => {
...
-            await postDocument(widget.id, {
-                title: data.name,
-                parent_document_id: data.folderId === 0 ? null : data.folderId,
-                url: data.link,
-                widget_id: widget.id,
-                type: 'file',
-            });
-            await loadDocuments();
+            if (documentToEdit) {
+                updateDocument(data);
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:94]: separate this into another method called saveDocument, have it like the example I shared earlier

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx [hunk 10]
```diff
@@ -140,10 +181,13 @@ const FileDrawer = () => {
                                     InputLabelProps={{
                                         shrink: false,
                                     }}
+                                    defaultValue={parentDocument?.title}
                                     fullWidth
                                     size="small"
                                 >
-                                    <MenuItem key={`folder-option-0`} value={0} sx={{ height: '2em' }}></MenuItem>
+                                    <MenuItem key={`folder-option-0`} value={0} sx={{ height: '2em' }}>
+                                        <i>none</i>
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:189]: just check if for italic there's a material UI prop in MenuItem, or maybe in the sx

File: met-web/src/services/widgetService/DocumentService.tsx/index.tsx [hunk 11]
```diff
@@ -32,3 +32,51 @@ export const postDocument = async (widget_id: number, data: PostDocumentRequest)
...
+    try {
+        const url = replaceAllInURL({
+            URL: Endpoints.Documents.DELETE,
+            params: {
+                document_id: String(document_id),
+                widget_id: String(widget_id),
+            },
+        });
+        const response = await http.DeleteRequest<DocumentItem>(url);
+        if (response.data.status && response.data.result) {
```
Comment (jadmsaadaot) [inline met-web/src/services/widgetService/DocumentService.tsx/index.tsx:46]: you could probably get away without this if statement. Perhaps lookup all the other delete service methods and maybe check if all of them have it

**Review by jadmsaadaot — APPROVED** [review]
few changes and questions

**Review by jadmsaadaot — CHANGES_REQUESTED** [review]
Few changes and a question

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx (hunk 7 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx:31]: I thought i removed that sorry about that that's weird

Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx:31]: I thought i removed that sorry about that that's weird removed.

File: met-web/src/services/widgetService/DocumentService.tsx/index.tsx (hunk 11 above)
Comment (djnunez-aot) [reply met-web/src/services/widgetService/DocumentService.tsx/index.tsx:46]: it seems it's the same in all of them

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx (hunk 10 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx:189]: i added it into fontStyle: 'italic'

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx (hunk 8 above)
Comment (djnunez-aot) [reply met-web/src/components/engagement/form/EngagementWidgets/Documents/FileDrawer.tsx]: I swear I changed these on my local cleaning that up now

File: met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx [hunk 12]
```diff
@@ -1,26 +1,124 @@
...
+        const updatedDocument = documents.find((document: DocumentItem) => document.id === documentItem.id);
+        setDocument(updatedDocument);
+    }, [documents]);
+
+    const updateDocument = async () => {
+        setEdit(false);
+        if (!documentWidget || !document) {
+            return;
+        }
+        if (document.title !== documentItem.title) {
```
Comment (jadmsaadaot) [inline met-web/src/components/engagement/form/EngagementWidgets/Documents/DocumentFolder.tsx:38]: What if there were many fields other than title? We could leverage updatedDiff before this block. updatedDiff will return an object with the differences. We perhaps can do something like Object.values(documentUpdatesToPatch ).length > 0. That way we are checking if there were any changes at all. We could perhaps  use that inside the patchDocument method itself and just return early if there's nothing to update

**Review by jadmsaadaot — CHANGES_REQUESTED** [review]
Thank you for the changes and the good work. I added one more comment

**Review by saravanpa-aot — APPROVED** [review]

**Pull request merged**